next-packages/provider/src/index.js

```diff
@@ -35,48 +30,24 @@ function getTheme(
     )
   }
 
-  if (outerTheme === undefined) {
-    return theme
-  }
-
   return { ...outerTheme, ...theme }
 }
 
-export default class Provider extends React.Component<Props> {
-  emotionCache: CSSContextType
-  consumer = (context: CSSContextType | null) => {
-    if (this.props.cache !== undefined) {
-      context = this.props.cache
+let createCreateCacheWithTheme = weakMemoize(cache => {
+  return weakMemoize(theme => {
+    let actualTheme = getTheme(cache.theme, theme)
+    return {
+      ...cache,
+      theme: actualTheme
     }
-    let newContext = context
-    if (context === null) {
-      if (this.emotionCache === undefined) {
-        newContext = createCache()
-        if (this.props.theme) {
-          newContext.theme = getTheme(this.props.theme, undefined)
-        }
-      } else if (
-        this.props.theme &&
-        this.props.theme !== this.emotionCache.theme
-      ) {
-        newContext = {
-          ...this.emotionCache,
-          theme: getTheme(this.props.theme, undefined)
-        }
-      }
-    } else if (this.props.theme && this.props.theme !== context.theme) {
-      newContext = {
-        ...context,
-        theme: getTheme(this.props.theme, context.theme)
-      }
-    }
-    return (
-      <CSSContext.Provider value={newContext}>
-        {this.props.children}
-      </CSSContext.Provider>
-    )
-  }
-  render() {
-    return <CSSContext.Consumer children={this.consumer} />
+  })
+})
+
+export default withCSSContext((props: Props, context) => {
```

Comment:
im slightly confused about those things:
- this package is called `provider` now but in fact it handles only theme and we now have things like `CSSContext.Provider`, not sure if generic `provider` name is a good one for this one here
- it exposes `withCSSContext`, but `@emotion/core` exposes it too (I realize their implementations differ but naming both the same is confusing)

Reply:
> this package is called provider now but in fact it handles only theme and we now have things like CSSContext.Provider, not sure if generic provider name is a good one for this one here

You're totally right about the naming. I'm not too focused on package naming right now since some stuff is likely going to move to other packages. For example, this component is likely going to be moved to emotion-theming and be called ThemeProvider.

> it exposes withCSSContext, but @emotion/core exposes it too (I realize their implementations differ but naming both the same is confusing)

I'm not sure if I'm understanding you correctly but this doesn't expose `withCSSContext`, it only uses it to define a component.

Reply:
> You're totally right about the naming. I'm not too focused on package naming right now since some stuff is likely going to move to other packages. For example, this component is likely going to be moved to emotion-theming and be called ThemeProvider.

👍 

> I'm not sure if I'm understanding you correctly but this doesn't expose withCSSContext, it only uses it to define a component.

Oh lol, i've misread this in a hurry 🙈